ex/loqui/lib/loqui/client.ex

```diff
@@ -0,0 +1,325 @@
+      registered_codecs: registered_codecs,
+      buffer: <<>>}
+
+    {:connect, :init, state}
+  end
+
+  def close(conn),
+    do: Connection.call(conn, :close)
+
+  def ping(conn) do
+    Connection.call(conn, :ping)
+  end
+
+  def request(conn, payload) do
+    Connection.call(conn, {:request, payload})
+  end
+
+  def push(conn, payload) do
+    Connection.cast(conn, {:push, payload})
+  end
+
+  def connect(_info, %{sock: nil, host: host, port: port, tcp_opts: tcp_opts, connect_timeout: connect_timeout}=state) do
+    with {:ok, sock}  <- :gen_tcp.connect(host, port, tcp_opts, connect_timeout),
+         {:ok, state} <- do_upgrade(%{state | sock: sock}),
+         state        <- make_active_once(state),
+         {:ok, state} <- do_loqui_connect(state) do
+
+      {:ok, state}
+    else
+      {:error, _} = error ->
+        Logger.error("Couldn't connect to #{host}:#{port} because #{inspect error}")
+        {:stop, error, state}
+    end
+  end
+
+  def disconnect(info, %State{sock: sock}=state) do
+    :ok = :gen_tcp.close(sock)
+    case info do
+
+      {:close, from} ->
+        Connection.reply(from, :ok)
+        {:stop, :normal, nil}
+
+      {:error, :closed} ->
+        Logger.error("Connection closed by server")
+        {:connect, info, %{state | sock: nil}}
+
+      {:error, reason} ->
+        reason = :inet.format_error(reason)
+        Logger.error("Connection error: #{inspect reason}")
+        {:connect, info, %{state | sock: nil}}
+    end
+  end
+
+  # Connection callbacks
+  def handle_call(:ping, caller, %{sock: sock}=state) do
+    {next_seq, state} = next_sequence(state)
+    :gen_tcp.send(sock, Frames.ping(0, next_seq))
+
+    {:noreply, %State{state | waiters: Map.put(state.waiters, next_seq, caller)}}
+  end
+
+  def handle_call(:close, from, %State{sock: sock}=s) do
+    go_away_packet = Frames.goaway(0, 0, "Closing")
+    :gen_tcp.send(sock, go_away_packet)
+
+    {:disconnect, {:close, from}, s}
+  end
+
+  def handle_call({:request, payload}, caller, %State{sock: sock, codec: codec}=state) do
+    {next_seq, state} = next_sequence(state)
+    encoded_payload = codec.encode(payload)
+    :gen_tcp.send(sock, Frames.request(0, next_seq, encoded_payload))
+
+    {:noreply, %State{state | waiters: Map.put(state.waiters, next_seq, caller)}}
+  end
+
+  def handle_cast({:push, payload}, %State{sock: sock, codec: codec}=state) do
+    encoded_payload = codec.encode(payload)
+    :gen_tcp.send(sock, Frames.push(0, encoded_payload))
+
+    {:noreply, state}
+  end
+
+  def handle_info({:tcp, _socket, data}, %State{}=state) do
+    with {:ok, parsed_packets, leftover_data} <- Parser.parse(state.buffer, data) do
+
+      state = handle_packets(parsed_packets, state)
+
+      {:noreply, %State{state | buffer: leftover_data}}
+    else
+
+      {:error, {:need_more_data, data}} ->
+        {:noreply, %State{buffer: data }}
+    end
+  end
+
+  def handle_info({:tcp_closed, _socket}, _state) do
```

Comment:
OS could send a error and get `{:tcp_error, _, _}` too. e.g. if someone used `tcp_opts: [show_econnreset: true]`